course6/toyapp-adam/math/vector.h

```diff
@@ -1,113 +1,172 @@
-	Vector2D operator-(const Vector2D& rhs) const { return {x - rhs.x, y - rhs.y}; }
-	Vector2D operator*(float scalar) const { return {x * scalar, y * scalar}; }
+	Vector()
+	:	mComponents{{0}}
+	{}
 
-	Vector2D& operator+=(const Vector2D& rhs)
+	template<typename... Args>
+	Vector(Args... args)
+	:	mComponents{{args...}}
+	{}
+
+	Vector(const Vector<Type, Dimensions-1>& other, Type lastComponent)
+	{
+		for (unsigned i = 0; i < Dimensions-1; ++i)
+			mComponents[i] = other[i];
+		mComponents[Dimensions-1] = lastComponent;
+	}
+
+	Vector(const Vector& other)
+	{
+		operator=(other);
+	}
+
+	operator Vector<Type, Dimensions-1>() const
+	{
+		Vector<Type, Dimensions-1> result;
+		for (unsigned i = 0; i < Dimensions-1; ++i)
+			result[i] = mComponents[i];
+		return result;
+	}
+
+	friend bool operator==(const Vector& lhs, const Vector& rhs)
 	{
-		x += rhs.x;
-		y += rhs.y;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			if (!ApproximatelyEquals(lhs[i], rhs[i]))
+				return false;
+		return true;
+	}
+
+	friend bool operator!=(const Vector& lhs, const Vector& rhs)
+	{
+		return !operator==(lhs, rhs);
+	}
+
+	Vector& operator=(const Vector& other)
+	{
+		for (unsigned i = 0; i < Dimensions; ++i)
+			mComponents[i] = other[i];
 		return *this;
 	}
 
-	Vector2D& operator-=(const Vector2D& rhs)
+	Vector& operator+=(const Vector& other)
 	{
-		x -= rhs.x;
-		y -= rhs.y;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			mComponents[i] += other[i];
 		return *this;
 	}
 
-	Vector2D& operator*=(float scalar)
+	Vector& operator*=(FloatType scalar)
 	{
-		x *= scalar;
-		y *= scalar;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			mComponents[i] *= scalar;
 		return *this;
 	}
 
-};
+	Vector& operator-=(const Vector& other)
+	{
+		return operator+=(-other);
+	}
 
-struct Vector3D
-{
-	float x = 0.0f;
-	float y = 0.0f;
-	float z = 0.0f;
+	Type& operator[](unsigned index)
+	{
+		assert(index < Dimensions);
+		return mComponents[index];
+	}
 
-	Vector3D operator-() const { return {-x, -y, -z}; }
-	Vector3D operator+(const Vector3D& rhs) const { return {x + rhs.x, y + rhs.y, z + rhs.z}; }
-	Vector3D operator-(const Vector3D& rhs) const { return {x - rhs.x, y - rhs.y, z - rhs.z}; }
-	Vector3D operator*(float scalar) const { return {x * scalar, y * scalar, z * scalar}; }
+	Type operator[](unsigned index) const
+	{
+		assert(index < Dimensions);
+		return mComponents[index];
+	}
 
-	Vector3D& operator+=(const Vector3D& rhs)
+	Vector operator-() const
 	{
-		x += rhs.x;
-		y += rhs.y;
-		z += rhs.z;
-		return *this;
+		Vector result;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			result[i] = -mComponents[i];
+		return result;
 	}
 
-	Vector3D& operator-=(const Vector3D& rhs)
+	Vector operator+(const Vector& rhs) const
 	{
-		x -= rhs.x;
-		y -= rhs.y;
-		z -= rhs.z;
-		return *this;
+		return Vector{*this} += rhs;
 	}
 
-	Vector3D& operator*=(float scalar)
+	Vector operator*(FloatType scalar) const
```

Comment:
Consider this simple form:
```
template<typename FloatType>
Vector operator*(FloatType scalar) const
{
   // ...
}
```
or, you you are feeling adventurous, you can either raise error if non-floating point type is used, lie that:
```
#include <type_traits>

template<typename FloatType>
Vector operator*(FloatType scalar) const
{
   static_assert(std::is_floating_point<FloatType>::value, "scalar must be float point value");
   // ...
}
```

or you can disable the function entirely for non-floating point types:

```
#include <type_traits>

template<typename FloatType, typename = std::enable_if_t<std::is_floating_point<FloatType>::value>>
Vector operator*(FloatType scalar) const
{
   // ...
}
```
The above is unreadable, you can make it easier by defining a helper:

```
#include <type_traits>
template<typename T>
using IsFloatingPointT = std::enable_if_t<std::is_floating_point<T>::value>;
```
then it will be easy to use, and readable:
```
template<typename FloatType, typename = IsFloatingPointT<FloatType>>
Vector operator*(FloatType scalar) const
{
   // ...
}
```